### ISSUE TITLE
loading last learning rate 

### ISSUE THREAD

**Issue by SamahZaro** [mozilla/DeepSpeech#2377]
Is it possible to add a feature to load the last reached learning rate value to continue the optimization from there?

Sometimes, basically when you have low resources, your training may stop unexpectedly!
Fine tuning from the last checkpoint or the best development on the same data and using the same learning rate will definitely give different results as if your stopped continued. This is due to learning rate change, right?

**Comment by lissyx** [comment]
> This is due to learning rate change, right?

hard to tell without more context ...


> Sometimes, basically when you have low resources, your training may stop unexpectedly!

That should not happen. But again, without more context ...


> Is it possible to add a feature to load the last reached learning rate value to continue the optimization from there?

I'm unsure about the need here: you have that in your shell script already

**Comment by SamahZaro** [comment]
Thank you for replying, @lissyx 

> > Sometimes, basically when you have low resources, your training may stop unexpectedly!
> 
> That should not happen. But again, without more context ...
> 
Ok, here is what I am having:
1. I have run a training on my own data with lr=0.0001.
2. It finished epoch 4 without any problems, this took around 16 hours on my low-resources laptop.
3. During epoch 5, the python process was killed unexpectedly due to RAM issues.
4. I need to continue training from where it left without loosing the progress. 
5. started new training with the same exact parameters, including lr=0.0001. Here, my model was not randomly initialized, but I loaded the best development checkpoint (which was saved after epoch 4) and doing fine tuning on it, again I need not to loose the past work. 
6. The loss results generated of steps from epoch 0 of the new training are different than those of epoch 5 from the killed past training.
7. I think they should be the same in my case. I think this is due to learning rate. starting from lr=0.0001, on epoch 5 the optimizer will work on decayed lr value not the 0.0001. right? 

> > Is it possible to add a feature to load the last reached learning rate value to continue the optimization from there?
> 
> I'm unsure about the need here: you have that in your shell script already

What I have in my shell script is the starting/initial learning rate. You are using AdamOptimizer, which updates the learning rate inside, step by step. The optimization of the next step depends on the this new learning rate value.

sorry for my English, I hope this is more clear now.

**Comment by reuben** [comment]
The learning rate and momentum variables of Adam are already saved in the checkpoint. The loss being different is most likely because epoch 0 of the new run is equivalent to starting epoch 5 from scratch again.

**Comment by SamahZaro** [comment]
>The loss being different is most likely because epoch 0 of the new run is equivalent to starting epoch 5 from scratch again.

epoch 0 of the new run should be equivalent to epoch 5 from scratch, but they are not. This is my question.

Thank you, @reuben .

**Comment by reuben** [comment]
Then it's something else. The learning rates are saved, but maybe something else is missing?

**Comment by kdavis-mozilla** [comment]
@SamahZaro any more info on this?

**Comment by lissyx** [comment]
@SamahZaro Gentle ping?

**Comment by lissyx** [comment]
Closing for lack of information and nothing actionable here.

**Comment by rhamnett** [comment]
I can confirm the learning rate is loaded correctly, as I submitted another patch to override learning rates that where set and stored by the reduce_lr_on_plateau
